src/nevermined/Nfts.ts

```diff
@@ -512,4 +515,148 @@ export class Nfts extends Instantiable {
             from
         )
     }
+
+    /**
+     *
+     * @param ddo {DDO} the Decentraized ID of the NFT
+     * @param assetRewards {AssetRewards} the currect setup of asset rewards
+     * @param nftAmount {Number} the number of NFTs put up for secondary sale
+     * @param provider {Account} the account that will be the provider of the secondary sale
+     * @param owner {Account} the account of the current owner
+     * @returns {Promise<string>} the agreementId if the secondary sale config was successful
+     */
+    public async listOnSecondaryMarkets(
+        ddo: DDO,
+        assetRewards: AssetRewards,
+        nftAmount: number,
+        provider: string,
+        token: Token,
+        owner: string
+    ): Promise<string> {
+        const { nftSalesTemplate } = this.nevermined.keeper.templates
+        const agreementId = zeroX(utils.generateId())
+        const nftSalesServiceAgreementTemplate = await nftSalesTemplate.getServiceAgreementTemplate()
+        const nftSalesTemplateConditions = await nftSalesTemplate.getServiceAgreementTemplateConditions()
+
+        nftSalesServiceAgreementTemplate.conditions = fillConditionsWithDDO(
+            nftSalesTemplateConditions,
+            ddo,
+            assetRewards,
+            token.getAddress(),
+            undefined,
+            provider || owner,
+            nftAmount
+        )
+
+        const nftSalesServiceAgreement: Service = {
+            type: 'nft-sales',
+            index: 6,
+            serviceEndpoint: this.nevermined.gateway.getNftEndpoint(),
+            templateId: nftSalesTemplate.getAddress(),
+            attributes: {
+                main: {
+                    name: 'nftSalesAgreement',
+                    creator: owner,
+                    datePublished: new Date().toISOString().replace(/\.[0-9]{3}/, ''),
+                    timeout: 86400
+                },
+                additionalInformation: {
+                    description: ''
+                },
+                serviceAgreementTemplate: nftSalesServiceAgreementTemplate
+            }
+        }
+
+        const saveResult = await this.nevermined.metadata.storeService(
+            agreementId,
+            nftSalesServiceAgreement
+        )
+
+        if (saveResult) {
+            return agreementId
+        } else {
+            throw Error(`Error saving ${agreementId} to MetadataDB`)
+        }
+    }
+
+    /**
+     * Buys a number of listed goods on secondary markets.
+     * @param buyer The address of the buyer.
+     * @param seller The address of the seller.
+     * @param numOfNfts The number of assets to buy. 1 by default.
+     * @param ddo The DDO of the NFT.
+     * @param provider The provider address.
+     * @param agreementId The agreementId of the initial sales agreement created off-chain.
+     * @returns true if the buy was successful.
+     */
+    public async buySecondaryMarketNft(
```

Comment:
I think this method just needs the buyer account, the agreementId and nftAmount. All the other information should be taken from the service agreement

Reply:
I added the provider just because of the potential of having a `delegate` do the mediating. Let's see if we can strip the others once the test cases succeed

Reply:
the provider is set during the did registry so it won't matter and the delegate uses nft level access control which is handled differently

Reply:
All the other information can and should be retrieved from the service agreement including the seller address

Reply:
I need to dig a bit deeper into these flows to understand them, right now it's a major blocker in my head. I'll try to take a look at them while on holidays 👍

Reply:
Params removed now 👍